tests/Events/SpanTest.php

```diff
@@ -42,4 +53,63 @@ public function testProducesValidJson(string $schemaVersion, string $schemaFile)
 
         $this->validateObjectAgainstSchema($span, $schemaVersion, $schemaFile);
     }
+
+    /**
+     * @throws \Nipwaayoni\Exception\Timer\AlreadyRunningException
+     *
+     * @dataProvider spanStartTimes
+     */
+    public function testUsesCorrectStartTime(float $startTime = null): void
+    {
+        /** @var EventBean|MockObject $parent */
+        $parent = $this->createMock(EventBean::class);
+        $parent->method('getId')->willReturn('123');
+        $parent->method('getTraceId')->willReturn('456');
+
+        $this->timer->expects($this->once())->method('stop');
+
+        $span = $this->makeTestSpan('MySpan', $parent, $startTime);
+        $span->start($startTime);
+        $span->stop();
+    }
+
+    public function spanStartTimes(): array
+    {
+        return [
+            'null start time' => [null],
+            'set start time' => [microtime(true)],
+        ];
+    }
+
+    private function makeTestSpan(string $name, EventBean $parent, float $expectedStart = null): Span
+    {
+        // The anonymous class does not have access to the members of the containing class, so give
+        // it a callable which will carry the necessary scope.
+        $timer = $this->timer;
```

Comment:
This is a good discussion to have. I'm still struggling with the veracity of the tests in this scenario. I have to assumptions.

1. The Span class is the SUT.
2. The expected creation of the Timer must be asserted.

Mocking Span::createTimer() requires that Span be mocked. Mocking the SUT is dangerous and counter to testing practicing.

Asserting behavior around the Timer creation requires more than just returning a timer. The object method assert method could be replaced with the static counterparts, but I don't think that addresses all the issues.

If the proper creation of the Timer is critical to the correct functioning of the Span class (which I believe it is at this time), then giving the Span a Factory it can use (and which can be easily mocked) allows us to address both concerns.

If I'm missing something in your suggestions that would allows us to accomplish both, I'm more than happy to learn more. I'm in the camp of "write less code", but I also have a strong desire to ensure tests are really valuable.

In the long run, I am going to propose the Span no longer create a Timer. Rather, it should inherit a Timer from its parent Transaction or Span. But that's a major redesign that is out-of-scope here.

Reply:
> This is a good discussion to have. I'm still struggling with the veracity of the tests in this scenario. I have to assumptions.

The Span class is the SUT.
The expected creation of the Timer must be asserted.
Mocking Span::createTimer() requires that Span be mocked. Mocking the SUT is dangerous and counter to testing practicing.

Agreed, although I used to see such approach on another libs. Using an anonymous class you decrease the degree of uncertainty because you are in the control. 

> If I'm missing something in your suggestions that would allows us to accomplish both, I'm more than happy to learn more. I'm in the camp of "write less code", but I also have a strong desire to ensure tests are really valuable.

No, you aren't missing anything. Actually this conversation is good to review some practices I used to use in my projects. 

> In the long run, I am going to propose the Span no longer create a Timer. Rather, it should inherit a Timer from its parent Transaction or Span. But that's a major redesign that is out-of-scope here.

Thinking coldly I have a question that could direct us in the best way to test Spans. Do we really need to test Span with `Timer` mocked? Maybe we only need to test if the duration is correctly after call stop. 

```php
$span = new Span();
$span->start();
sleep(1);
$span->stop();
$this->assertEquals(1000, $span->getDuration();
```

Reply:
This where I start to think the Span should not be doing its own timing.  According to the spec, a span takes a timestamp or a start (offset relative to the transaction's timestamp) and the duration. The current Span object provides timestamp and duration. Rather than adding a `getDuration()` method, we could convert the Span to the data model after stopping it and validate those two values.

However, trying to right tests around real time is going to be a problem. I tried that with the Timer class several months ago and ended up having to just expect a value greater than the sleep time. The cycles to even execute the sleep and stop methods can throw the timing off and I even with that I sometimes see those tests fail.

This was one reason I proposed injecting our own Timer object. That would put us in complete control and able to return exactly the duration we want, removing the variance of actual microsecond timing during tests. As long as Span depends on a Timer, taking control of the mechanism for creating the Timer allows us to assert the Span creates it correctly and also return fixed values to the Span.

I agree with your point about not creating abstractions which are not needed. When a class starts taking a lot of effort to test, though, I think that might be a reason to consider abstractions or other design choices.

Reply:
I find dates are relatively easy if you consistently use `Carbon` since it lets you set `now` during testing. We're dealing with timing of events at a microsecond scale and there is just too much variance between executions for that to be precise in tests. We should treat time as an external system, isolate it in the smallest possible implementation and mock when testing classes which depend on it. I think we're getting closer, though.

Reply:
I think this conversation is addressed by the recently merged `TimerFactory` implementation. While we still have a lot of improvements to make on the event representation and timing implementation, this should move us forward.

Reply:
I believe this is resolved. I'm going to close this conversation so we can proceed with the merge.